app/validators/questionnaire_schema.py

```diff
@@ -0,0 +1,298 @@
+                            **context,
+                        }
+
+        return answers
+
+    @cached_property
+    def is_hub_enabled(self):
+        return self.schema.get("hub", {}).get("enabled")
+
+    @cached_property
+    def ids(self):
+        """
+        question_id & answer_id should be globally unique with some exceptions:
+            - within a block, ids can be duplicated across variants, but must still be unique outside of the block.
+            - answer_ids must be duplicated across add / edit blocks on list collectors which populate the same list.
+        """
+        unique_ids_per_block = defaultdict(set)
+        non_block_ids = []
+        all_ids = []
+
+        for path, value in self.id_paths:
+            if "blocks" in path:
+                # Generate a string path and add it to the set representing the ids in that path
+                path_list = path.split(".")
+
+                block_path = path_list[: path_list.index("blocks") + 2]
+
+                string_path = ".".join(block_path)
+                # Since unique_ids_per_block is a set, duplicate ids will only be recorded once within the block.
+                unique_ids_per_block[string_path].add(value)
+            else:
+                non_block_ids.append(value)
+
+        for block_ids in unique_ids_per_block.values():
+            all_ids.extend(block_ids)
+        all_ids.extend(non_block_ids)
+
+        return all_ids
+
+    @cached_property
+    def id_paths(self):
+        """
+        These values will be returned with the json path to them through the object e.g.
+            - 'sections.[0].groups[0].blocks[1].question_variants[0].question.question-2'
+
+        Returns: generator yielding (path, value) tuples
+        """
+        ignored_paths = [
+            "routing_rules",
+            "skip_conditions",
+            "when",
+            "edit_block.question.answers",
+            "add_block.question.answers",
+            "remove_block.question.answers",
+            "edit_block.question_variants.answers",
+            "add_block.question_variants.answers",
+            "remove_block.question_variants.answers",
+        ]
+
+        for match in parse("$..id").find(self.schema):
+            full_path = str(match.full_path)
+            if not any(ignored_path in full_path for ignored_path in ignored_paths):
+                yield str(match.full_path)[:-3], match.value
+
+    @cached_property
+    def answer_id_to_option_values_map(self):
+        answer_id_to_option_values_map = defaultdict(set)
+
+        for answer in self.answers:
+            if "options" not in answer:
+                continue
+
+            answer_id = answer["id"]
+            option_values = [option["value"] for option in answer["options"]]
+
+            answer_id_to_option_values_map[answer_id].update(option_values)
+
+        return answer_id_to_option_values_map
+
+    @cached_property
+    def answers(self):
```

Comment:
I think some of these caching between schemas. I get empty generators, if I remove caching then it actually returns them.

Reply:
The only reason these would be caching between schemas is that the QuestionnaireSchema is reused across many tests. As far as I'm aware the tests shouldn't be structured in this way.

Reply:
Sorry I am referring to falsely passing schemas when validating them. The unit tests run fine, but I ran into the issue when running against the schemas in runner for example.